src/pmi/pmi.cs

```diff
@@ -100,85 +398,323 @@ private static TimeSpan PrepareMethod(Type t, MicroMethodInfo mMI)
         }
 
-        for (int i = 0; i < ci.Length; i++)
+        // Types we will use for instantiation attempts.
+        Type[] typesToTry = new Type[] { typeof(object), typeof(int), typeof(double), typeof(Vector<float>) };
+
+        // To keep things sane, we won't try and instantiate too many copies
+        int instantiationLimit = genericArguments.Length * typesToTry.Length;
+        int instantiationCount = 0;
+
+        foreach (Type firstType in typesToTry)
         {
-            mMI[i + mi.Length].RMH = ci[i].MethodHandle;
-            mMI[i + mi.Length].name = ci[i].Name;
-            mMI[i + mi.Length].IsAbstract = ci[i].IsAbstract;
-            mMI[i + mi.Length].ContainsGenericParameters = ci[i].ContainsGenericParameters;
+            bool areConstraintsSatisfied = AreConstraintsSatisfied(firstType, genericArguments[0]);
+
+            Type secondType = null;
+
+            if (genericArguments.Length == 2)
+            {
+                foreach (Type secondTypeX in typesToTry)
+                {
+                    areConstraintsSatisfied &= AreConstraintsSatisfied(secondTypeX, genericArguments[1]);
+                    secondType = secondTypeX;
+                }
+            }
+
+            if (!areConstraintsSatisfied)
+            {
+                continue;
+            }
+
+            // Now try and instantiate.
+            try
+            {
+                Type newType = null;
+
+                if (genericArguments.Length == 1)
+                {
+                    newType = type.MakeGenericType(firstType);
+                }
+                else if (genericArguments.Length == 2)
+                {
+                    newType = type.MakeGenericType(firstType, secondType);
+                }
+
+                // If we can instantiate, prepare the methods.
+                if (newType != null)
+                {
+                    results.Add(newType);
+                }
+            }
+            catch (Exception)
+            {
+                // Probably missing a constraint check
+            }
+
+            if (instantiationCount >= instantiationLimit)
+            {
+                break;
+            }
         }
 
-        return mMI;
+        return results;
     }
 
-    private static void WriteAndFlushNextMethodToPrepMarker(int methodBeingPrepped)
+    // Try and identify obviously invalid type substitutions.
+    //
+    // It is ok if we miss some, as we catch the exception that will
+    // arise when instantating with an invalid type.
+    static bool AreConstraintsSatisfied(Type type, Type parameterType)
     {
-        int nextMethodToPrep = (methodBeingPrepped + 1);
+        bool areConstraintsSatisfied = true;
+        GenericParameterAttributes gpa = parameterType.GenericParameterAttributes;
 
-        using (var writer = new StreamWriter(File.Create("NextMethodToPrep.marker")))
+        if ((gpa & GenericParameterAttributes.ReferenceTypeConstraint) != 0)
```

Comment:
I'm probably missing something here, but if this condition is true, you won't check for the other constraints below. I'm not sure if `ReferenceTypeConstraint` would be "optimized away" on the type if it specifies a base type, but I would think that an interface constraint could still apply.

Reply:
I think you're right -- if this check succeeds we still should do the other checks.

It's ok for this method to be a bit sloppy, The runtime will do the real checking when PMI tries to instantiate the type and will throw an exception if the types aren't suitable. 

But it's nice if we can catch these up front to reduce the volume of unnecessary exceptions.

Reply:
Thanks